File: LeetCode/valid-parentheses.cpp
// https://leetcode.com/problems/valid-parentheses/

class Solution
{
public:
	bool isValid(string s)
	{
		if (s[0] == ')' || s[0] == '}' || s[0] == ']' || s.size() == 1)
		{
			return false;
		}
		else
		{
			map<char, char> m;
			map<char, char>::iterator itr;
			m.insert({')', '('});
			m.insert({'}', '{'});
			m.insert({']', '['});

			stack<char> st;

			for (int i = 0; i < s.size(); i++)
			{
				itr = m.find(s[i]);
				if (itr != m.end() && !st.empty())
				{
					if (itr->second == st.top())
					{
						st.pop();
					}
					else
					{
						return false;
					}
				}
				else
				{
					st.push(s[i]);
				}
			}
			if (st.empty())
			{
				return true;
			}
			else
			{
				return false;
			}
		}
	}
};

File: LeetCode/Java/validparentheses/ValidParentheses.java
package validparentheses;

import java.util.Stack;

public class ValidParentheses {
	public static void main(String[] args) {
		Solution sol = new Solution();
		boolean res = sol.isValid("[");
		System.out.println(res);
	}
}

// https://leetcode.com/problems/valid-parentheses/

//IMPORTANT!! Submit Code Region Begin(Do not remove this line)
class Solution {
	public boolean isValid(String s) {
		Stack<Character> stack = new Stack<>();

		// push expected closing brackets for each opening bracket.
		// if current char is a closing bracket, it must match the top of the stack.
		for (char ch : s.toCharArray()) {
			if (ch == '(') {
				stack.push(')');
			} else if (ch == '{') {
				stack.push('}');
			} else if (ch == '[') {
				stack.push(']');
			} else if (stack.isEmpty() || stack.pop() != ch) {
				return false;
			}
		}

		return stack.isEmpty();
	}
}
//IMPORTANT!! Submit Code Region End(Do not remove this line)